jax/lax/lax_control_flow.py

```diff
@@ -888,65 +893,81 @@ def _cond_partial_eval(trace, *tracers, branches, linear):
 # staged jaxpr that accepts those residuals as its first few inputs. The
 # residual-producing branches are staged as jaxprs and bound right away in a
 # conditional. The residual-consuming jaxprs are assembled together in a jaxpr
-# conditional. The following two helper functions are used to ensure that both
-# collections of jaxprs (those evaluated and those staged) are valid for joint
-# use under their respective conditionals.
-
-# Because every branch might produce different residuals, the branches' output
-# signatures might not match. But we need branch signatures to match in order to
-# bind them in a conditional. This function "joins" the residual outputs of the
-# branches by concatenation. Each augmented branch returns zero-filled values in
-# the place of all other branches' residuals.
-def _join_cond_outputs(jaxprs, res_avals_per_jaxpr, num_non_res_outputs):
-  def augment_jaxpr(i, jaxpr):
-    res_avals_prefix = util.concatenate(res_avals_per_jaxpr[:i])
-    res_avals_suffix = util.concatenate(res_avals_per_jaxpr[i+1:])
-
+# conditional. The following helper functions ensure that both collections of
+# jaxprs (those evaluated and those staged) are valid for joint use under their
+# respective conditionals.
+#
+# In particular, the residuals derived from each original branch may have
+# distinct types. Because the branches of conditionals must have identical type
+# signatures, we join residuals together across branches into a common format.
+
+# In order to set up a type signature that all branches can conform to, it would
+# suffice to concatenate all branches' residuals. But concatenation can result
+# in redundant inputs and outputs, and might result in memory allocation that
+# scales unnecessarily with the branch count. This function finds common
+# residual types across branches for reuse, so as to avoid redundant allocation.
+def _merge_branch_residuals(branch_res_avals):
+  def enumerate_equal(xs):
+    counts = {v: itertools.count() for v in set(xs)}
+    return [(x, next(counts[x])) for x in xs]
+  branch_res_avals = _map(enumerate_equal, branch_res_avals)
+  all_avals = sorted(set(util.concatenate(branch_res_avals)))
+  all_avals_indices = {v: i for i, v in enumerate(all_avals)}
+  branch_indices = [
+      [all_avals_indices[aval] for aval in avals] for avals in branch_res_avals]
```

Comment:
IIUC `avals` here is a list of (aval, int) pairs, and each `aval` is an (aval, int) pair, so I wonder if different names could be chosen.

To continue the narration of the logic of this function: once we have the union-of-disjoint-unions which has computed the deduped set of residuals needed, we need to match these residual slots back up with each branch's residuals. That is, we need an index. We set up index values for each of the deduped residuals and a mapping from (aval, int) pair (i.e. the disjoint-union-tagged aval) to those indices. Then we just need to walk those disjoint unions we computed earlier and produce the index value for each element. The final result is we have a list-of-list of ints of the same lengths as the list-of-lists of avals we got as input, where each int tells us which deduped residual is responsible for the corresponding branch's residual.

Super slick!

Reply:
:) Your reading is correct. I've changed some names, hopefully making this easier to follow.